nav2_collision_monitor/src/costmap.cpp

```diff
@@ -0,0 +1,149 @@
+#include <nav2_ros_common/lifecycle_node.hpp>
+#include <nav2_ros_common/node_utils.hpp>
+
+namespace nav2_collision_monitor
+{
+
+
+CostmapSource::CostmapSource(
+  const nav2::LifecycleNode::WeakPtr & node,
+  const std::string & source_name,
+  const std::shared_ptr<tf2_ros::Buffer> tf_buffer,
+  const std::string & base_frame_id,
+  const std::string & global_frame_id,
+  const tf2::Duration & transform_tolerance,
+  const rclcpp::Duration & source_timeout,
+  const bool base_shift_correction)
+: Source(
+    node, source_name, tf_buffer, base_frame_id, global_frame_id,
+    transform_tolerance, source_timeout, base_shift_correction),
+  data_(nullptr)
+{
+  RCLCPP_INFO(logger_, "[%s]: Creating CostmapSource", source_name_.c_str());
+}
+
+
+CostmapSource::~CostmapSource()
+{
+  RCLCPP_INFO(logger_, "[%s]: Destroying CostmapSource", source_name_.c_str());
+  data_sub_.reset();
+}
+
+void CostmapSource::configure()
+{
+  Source::configure();
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+  std::string source_topic;
+  getParameters(source_topic);
+  rclcpp::QoS qos = rclcpp::SystemDefaultsQoS();
+
+
+  data_sub_ = node->create_subscription<nav2_msgs::msg::Costmap>(
+      source_topic,
+     std::bind(&CostmapSource::dataCallback, this, std::placeholders::_1),
+      qos);
+}
+
+bool CostmapSource::getData(
+  const rclcpp::Time & curr_time,
+  std::vector<Point> & data)
+{
+  if (data_ == nullptr) {
+    return false;
+  }
+
+  if (!sourceValid(data_->header.stamp, curr_time)) {
+    return false;
+  }
+  tf2::Transform tf_transform; tf_transform.setIdentity();
+  const std::string src = data_->header.frame_id;
+  if (src != base_frame_id_) {
+    if (!getTransform(curr_time, data_->header, tf_transform)) {
+      RCLCPP_WARN(logger_, "[%s] TF %s->%s unavailable at t=%.3f",
+        source_name_.c_str(), src.c_str(), base_frame_id_.c_str(), curr_time.seconds());
+      return false;
+    }
+  }
+
+
+  // Extract lethal/inscribed cells and transform to base frame
+  const auto & cm = *data_;
+  const auto & meta = cm.metadata;
+
+  for (unsigned int y = 0; y < meta.size_y; ++y) {
+    for (unsigned int x = 0; x < meta.size_x; ++x) {
+      const int idx = y * meta.size_x + x;
+
+      const uint8_t c = cm.data[idx];
+      const bool is_obstacle = (c >= cost_threshold_ && c < 255) ||
+        (treat_unknown_as_obstacle_ && c == 255);
+      if (is_obstacle) {
+        const double wx = meta.origin.position.x + (x + 0.5) * meta.resolution;
+        const double wy = meta.origin.position.y + (y + 0.5) * meta.resolution;
+        tf2::Vector3 p_v3_s(wx, wy, 0.0);
+        tf2::Vector3 p_v3_b = tf_transform * p_v3_s;
+        data.push_back({p_v3_b.x(), p_v3_b.y()});
+        RCLCPP_INFO(logger_, "[%s] Lethal cell at (%f, %f)",
+          source_name_.c_str(), wx, wy);
+      }
+    }
+  }
+  return true;
+}
+
+void CostmapSource::getParameters(std::string & source_topic)
```

Comment:
All params should be added to  the collision monitor's configuration guide at docs.nav2.org